pczt/src/roles/signer/mod.rs

```diff
@@ -0,0 +1,302 @@
+    sighash_v5::v5_signature_hash,
+    txid::TxIdDigester,
+    Authorization, TransactionData, TxDigests, TxVersion,
+};
+use zcash_protocol::consensus::BranchId;
+
+use crate::{common::Global, Pczt};
+
+use super::tx_extractor::determine_lock_time;
+
+const V5_TX_VERSION: u32 = 5;
+const V5_VERSION_GROUP_ID: u32 = 0x26A7270A;
+
+pub struct Signer {
+    global: Global,
+    transparent: transparent::pczt::Bundle,
+    sapling: sapling::pczt::Bundle,
+    orchard: orchard::pczt::Bundle,
+    /// Cached across multiple signatures.
+    tx_data: TransactionData<EffectsOnly>,
+    txid_parts: TxDigests<Blake2bHash>,
+    shielded_sighash: [u8; 32],
+    secp: secp256k1::Secp256k1<secp256k1::SignOnly>,
+}
+
+impl Signer {
+    /// Instantiates the Signer role with the given PCZT.
+    pub fn new(pczt: Pczt) -> Result<Self, Error> {
+        let Pczt {
+            global,
+            transparent,
+            sapling,
+            orchard,
+        } = pczt;
+
+        let transparent = transparent.into_parsed().map_err(Error::TransparentParse)?;
+        let sapling = sapling.into_parsed().map_err(Error::SaplingParse)?;
+        let orchard = orchard.into_parsed().map_err(Error::OrchardParse)?;
+
+        let tx_data = pczt_to_tx_data(&global, &transparent, &sapling, &orchard)?;
+        let txid_parts = tx_data.digest(TxIdDigester);
+
+        // TODO: Pick sighash based on tx version.
+        let shielded_sighash = v5_signature_hash(&tx_data, &SignableInput::Shielded, &txid_parts)
+            .as_ref()
+            .try_into()
+            .expect("correct length");
+
+        Ok(Self {
+            global,
+            transparent,
+            sapling,
+            orchard,
+            tx_data,
+            txid_parts,
+            shielded_sighash,
+            secp: secp256k1::Secp256k1::signing_only(),
+        })
+    }
+
+    /// Signs the transparent spend at the given index with the given spending key.
+    ///
+    /// It is the caller's responsibility to perform any semantic validity checks on the
+    /// PCZT (for example, comfirming that the change amounts are correct) before calling
+    /// this method.
+    pub fn sign_transparent(
+        &mut self,
+        index: usize,
+        sk: &secp256k1::SecretKey,
+    ) -> Result<(), Error> {
+        let input = self
+            .transparent
+            .inputs_mut()
+            .get_mut(index)
+            .ok_or(Error::InvalidIndex)?;
+
+        // Check consistency of the input being signed.
+        // TODO
+
+        input
+            .sign(index, &self.tx_data, &self.txid_parts, sk, &self.secp)
+            .map_err(Error::TransparentSign)?;
+
+        // Update transaction modifiability.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !0b0000_0001;
+        }
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
```

Comment:
If (as above) it is intended that the modifiability flags only prevent modifications that are definitely incorrect, then the outputs should only be made non-modifiable for `SighashType::{ALL, ALL_ANYONECANPAY}`. In the case of `SighashType::{SINGLE, SINGLE_ANYONECANPAY}`, other outputs than the one corresponding to the input being signed can still be modified.
```suggestion
        if (input.sighash_type() & !SIGHASH_ANYONECANPAY) == SighashType::ALL {
```

Reference for the semantics of sighash types: https://developer.bitcoin.org/devguide/transactions.html#signature-hash-types

Reply:
The semantics of `tx_modifiable` come directly from [BIP 370](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#signer) (formatting added for clarity):
> For PSBTv2s, a signer must update the PSBT_GLOBAL_TX_MODIFIABLE field after signing inputs so that it accurately reflects the state of the PSBT.
> - If the Signer added a signature that does not use SIGHASH_ANYONECANPAY, the Input Modifiable flag must be set to False.
> - If the Signer added a signature that does not use SIGHASH_NONE, the Outputs Modifiable flag must be set to False.
> - If the Signer added a signature that uses SIGHASH_SINGLE, the Has SIGHASH_SINGLE flag must be set to True. 

Technically someone can add an output after an existing SIGHASH_SINGLE-constrained one, but given that we want to match BIP 370 semantics where possible for ease of usage alongside PSBTs, I'm fine with this being more restrictive.